pages/articles/2018-01-31-Neural-Program-Search-Paper/index.md

```diff
@@ -0,0 +1,47 @@
+---
+layout: post
+title: Neural Program Search Paper
+date: "2018-01-31T11:51:03.284Z"
+---
+
+We are excited to introduce our paper "Neural Program Search - Solving Programming Tasks From Description and Examples".
+
+There are two main approaches currently to program synthesis: from some form of specification and from examples of Inputs / Outputs.
+First type, can vary from natural language descrition of the problem, to formal First-Order Logic expression that describes what program must achieve.
+First-order logic (or it's forms in Prolog) require quiet a different way of thinking to describe a problem at hand, and thus are not very practical.
+Natural language on the other hand can be quiet unclear or ambiguous about some details of calcualtion.
+
+Second type - from IO examples, has received more success in practice (in works such as FlashFill), for some of the domains 
+of programming (specifically string manipulation) provides a very clear way of specifying what program must do.
+But either requires a large number of examples or very restricted program space to make it work.
+
+On the other hand, there are two methods of addressing program synthesis, 
+ * Classic method via searching programs using heuristics in exhausive search algorithm
+ * More recent methods of applying deep learning to producing programs via sequence or similar decoding
+
+In this work, we are introducing a method that combines both ways of specifying problems as well as combination of methods to solve them.
+
+<!--more-->
+Deep Learning has made progress at natural language understanding with recent improvements in semantic parsing, machine translation and etc.
+But it's still problematic to produce specific structure, as neural networks just find correlations between inputs. (??????)
```

Comment:
"However, due to the probabilistic nature of neural networks, they do not work well for code synthesis, where one misplaced token can completely change the meaning of the program"?